PDF Inventory where title is PMID

quang-ng opened the pull request:
Work for issue #22

quang-ng committed:
Implement S3 inventory processing and add unit tests for OddpubWrapper

quang-ng commented:
the file changes look like crazy now.!!!! Create new branch for this ticket.

quang-ng committed:
Remove irp_pdf.py and refactor S3 inventory processing into UploadPDFsTitleIsPMID class. Add unit tests for OddpubWrapper to ensure PDF processing and S3 inventory functionality. Enhance error handling and logging.

quang-ng committed:
Integrate Oddpub analysis with error handling and logging for improved robustness.

quang-ng committed:
Add command-line interface for PDF upload script

quang-ng committed:
Add unit tests for UploadPDFsTitleIsPMID class. Remove obsolete test file for OddpubWrapper. Enhance error handling and logging in S3 inventory processing.

quang-ng committed:
Refactor UploadPDFsTitleIsPMID class

quang-ng committed:
Refactor PDF upload processing method in UploadPDFsTitleIsPMID class

leej3 commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,160 @@
+
+
+def main():
+    parser = argparse.ArgumentParser(
```
I think moving cli parsing logic to the scripts directory would be good.

leej3 commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,160 @@
+        required=True,
+        help="The database connection URL. This should be a valid SQLAlchemy database URL.",
+    )
+    parser.add_argument(
```
> 3538 pdfs are in the osm-pdf-uploads bucket

The issue specifies the above input. Perhaps the code should be able to take an s3 url or a local path? @joshlawrimore any preferences?

quang-ng replied:
I removed this argument, we only get pdfs files from predefine bucket

joshlawrimore replied:
Both would be preferable given my fondness for local user testing

leej3 commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,160 @@
+        self.s3_client = boto3.client("s3")
+        self.oddpub_host_api = oddpub_host_api
+
+    def process_s3_bucket(self) -> str:
```
I think this method could be named better. Adding a docstring might facilitate the naming. Perhaps just `run`?

joshlawrimore replied:
You seem to be doing two things here, iterating over a bucket's contents for inventory and running oddpub. You could name it
```suggestion
    def pmid_inventory_oddpub(self) -> str:
```

leej3 commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,160 @@
+        except Exception as e:
+            logger.error(f"Error processing S3 inventory: {str(e)}")
+
+    def _get_s3_page_iterator(self):
```
This method could be named better. Is this just listing objects in the bucket? If so, `_list_s3_objects` would be appropriate. If you are indeed using pagination to batch the list of objects I would prefer the term "batch" instead. I think "page" is overloaded here because one can have pages of an article... so `_get_s3_object_batches` but I don't fully understand this code.

quang-ng replied:
yes, the function use to fetch pdf files by batch.

leej3 commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,160 @@
+
+    def process_s3_bucket(self) -> str:
+        try:
+            page_iterator = self._get_s3_page_iterator()
```
As this is the core logic of the class it would be helpful to add comment as to why things are being done here. It's not clear to me what a "page" is. Try to adjust the variable names to the task being done:
we want to iterate through pdfs, extract metrics for them, upload them, upload the metrics.

I don't understand why the existing hashes functionality is here. My understanding is that checking for pre-existing hashes doesn't help because:
- We don't yet care about document duplication
- Document dupllication can occur by running different processing pipelines (oddpub won't be the only metric extraction tool we use)
- Even if documents are duplicates we likely won't even detect that from the hash because of publication source, publication version, water-marking, and other metadata injected into the pdf upon download.

leej3 replied:
Some of the above may not be relevant after a reorganisation. And it seems like "batch" should replace "page"... I think that address that confusion upon initial reading of the code.

leej3 commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,160 @@
+
+    def _create_provenance_entry(self):
+        provenance = Provenance(
+            pipeline_name="Document Upload",
```
This is running oddpub on the pdf documents... that should be the name of the pipeline. We will be running other tools in the future though see my other note regarding the granularity at which we specify provenance entries.

leej3 commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,160 @@
+            document = Documents(
+                hash_data=file_hash,
+                s3uri=f"s3://{self.bucket_name}/{key}",
+                provenance_id=provenance.id,
```
@joshlawrimore should each of these documents have it's own provenance entry?
I think the options would be:
- it would seem that a single script execution could describe the upload/oddpub processing for all entries in the Document table  and the Rtransparent/Oddpub/Processing_X table. 
The schema may already imply the answer here so apologies if that is so! My understanding was that we would have a provenance table entry for every operation so:
An entry for upload, and oddpub execution for every pdf.

joshlawrimore replied:
Correct. A single provanance id corresponding to the inventory/oddpub analysis should be applies to all the document table entries. I haven't run this code locally yet, but I think that is what should happen as this is writen.

leej3 commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,160 @@
+                provenance_id=provenance.id,
+            )
+            self.db_session.add(identifier)
+            self.db_session.commit()
```
we likely want to make all interaction with the database atomic i.e. for any given document either the full processing succeeds and all tables have the entry inserted or nothing is added.

joshlawrimore replied:
Yes please. I :heart atomicity.

leej3 commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,160 @@
+from .config import config
+
+
+class UploadPDFsTitleIsPMID:
```
I'm not sure about the choice of a class here. Much of this functionality is shared with pre-existing scripts. Ideally we would reuse that functionality, making it more generalisable where necessary to fit our purposes.
It may be the appropriate level of encapsulation though so feel free to explain the pros/cons of this approach vs another.

joshlawrimore replied:
```suggestion
class DocumentInventoryPMID:
```

joshlawrimore replied:
You are not really uploading anything. I used the wrong word [Issue 22](https://github.com/nimh-dsst/dsst-etl/issues/22). The idea is to create a document inventory where the identifer the is PMID in the title.

leej3 reviewed (changes requested):
Thanks @quang-ng. It looks like it is coming along nicely. There  are a couple of clarifications on the objective I have requested from @joshlawrimore.

Overall I think it's almost there. I would prefer to see more code reuse though. PDF uploads and the associated population of tables has already done. Hopefully we can reuse that.

quang-ng committed:
Move main method to separate script and simplify PDF upload processing

quang-ng committed:
Rename process_s3_bucket method to run and add comprehensive docstring

quang-ng committed:
Replace page-based iteration with batch processing of PDF files.
Removing duplicate hash checks

quang-ng committed:
Added rollback mechanism for database operations to ensure data integrity during PDF uploads.

quang-ng committed:
fix unit-test

quang-ng committed:
fix unit-test

quang-ng committed:
Update pipeline name

quang-ng committed:
OddpubWrapper to use log

quang-ng committed:
Added a new method for Oddpub analysis, refined transaction rollback logic, and removed redundant logging.

joshlawrimore commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,162 @@
+
+class UploadPDFsTitleIsPMID:
+    """
+    Uploads PDFs to S3 where the title is the PMID.
```
```suggestion
    Inventory PDFs to S3 where the title is the PMID.
```

joshlawrimore commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,162 @@
+            version=__version__,
+            compute=get_compute_context_id(),
+            personnel=config.HOSTNAME,
+            comment="Upload PDFs where the title is the PMID",
```
```suggestion
            comment="Creating document inventory of PDFs where the title is the PMID",
```

joshlawrimore commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,162 @@
+    def _create_provenance_entry(self):
+        # Creates a provenance entry to track the current upload process
+        provenance = Provenance(
+            pipeline_name="Oddpub Analysis",
```
```suggestion
            pipeline_name="PMID Inventory and Oddpub Analysis",
```

joshlawrimore commented:
Please let me know if I missed any questions. I am going to attempt to run the code on a subset of the pdfs in the `osm-pdf-uploads` bucket and pipe the results to a local postgres database. I will let you know how that goes.

@quang-ng , when do you take vacation?

joshlawrimore commented on dsst_etl/upload_pdfs_title_is_pmid.py:
```diff
@@ -0,0 +1,162 @@
+
+        This method performs the following steps:
+        1. Retrieves an iterator for paginated S3 objects.
+        2. Creates a provenance entry for the current upload process.
```
```suggestion
        2. Creates a provenance entry for the current inventory process.
```

leej3 committed:
Apply some renaming for clarity

Co-authored-by: joshlawrimore <86373443+joshlawrimore@users.noreply.github.com>

joshlawrimore commented:
The way the oddpub_metrics table's article column is defined:
[op.create_index(op.f('ix_oddpub_metrics_article'), 'oddpub_metrics', ['article'], unique=True)](https://github.com/nimh-dsst/dsst-etl/blob/99dd9d9d71e2981f03b4ec17e5d3faf7d70e2bd3/alembic/versions/832c238c1be7_add_oddpub_metrics_table.py#L44-L45)
states that the value of the article must be unique. I don't think this is a good idea for two reasons:

1. When the file contents of a PDF is pulled from an S3_bucket, the default value for the article is `file`. When the file contents of a PDF is pulled from a local file, the article is the name of the PDF with a `.txt` file extension. That means that the code written in [upload_pdfs_title_is_pmid.py](https://github.com/nimh-dsst/dsst-etl/blob/irp-pdf/dsst_etl/upload_pdfs_title_is_pmid.py) does not work on my local machine. I didn't spin up an EC2 cluster to test it on AWS, so it may work there.
2. Even if the code does work, I think we should leave the article column value as non-unique as it represents the results of the how the code was run. No need to specify uniqueness at this level.

@leej3 thoughts?

leej3 committed:
fix tests

leej3 committed:
tidy docker compose

leej3 committed:
tidy readme

leej3 committed:
include minio in testing

leej3 commented:
> Both would be preferable given my fondness for local user testing
> ...
> @leej3 thoughts?

I agree with supporting local execution to  help develop/review. I think adding logic for interaction with files on the file system should be minimised though if that is not a natural part of the pipeline or required interface for the user. My understanding is that here it is not a natural part of the interface. There's a lot of gotchas that need to be worked through that aren't fulfilling the objective of the tasks. There is not anything too arduous but adding cli options, adding additional control flow to account for two different execution streams for every script adds up.

There are various tools to emulate an s3 API that can be deployed by docker. I think that might be a nicer approach. That way we are more fully testing the functionality that interacts with s3 but we don't need to actually deploy the AWS resources for testing and resetting state is as simple as `docker-compose down -v`. I've pushed a prototype of what that would look like.

I've also separated the services into ones that are required by scripts/pipelines (compose.yaml) and those that are accumulating as development dependencies like postgres and minio (compose.override.yaml). The override file is used by default. Explicitly running the base file: `docker compose -f compose.yaml up` will just run the core services.  

> When the file contents of a PDF is pulled from a local file, the article is the name of the PDF with a .txt file extension.

This would be the sort of additional logic that the above would avoid.

> No need to specify uniqueness at this level.

Agreed.

leej3 committed:
update ci

joshlawrimore committed:
Rename dockerfile to Dockerfile

Podman does not like it when the cases do not match. Git doesn't seem to care about changing case for filenames for some reason... changing it now...

leej3 committed:
use "inventory" instead of "upload"

joshlawrimore committed:
Alembic revisions to merge heads and to prevent failure due to non-unique article values in oddpub_metrics table. Added podman-compose package to dev for Josh's local testing.

joshlawrimore committed:
Updated article column in OddpubMetrics class to remove index and unique parameters. Makes definition compatilbe with alembic revisions

joshlawrimore committed:
Added podman compose installation and setup instructions to README.md

joshlawrimore committed:
AI generated apptainer definition for oddpub

joshlawrimore committed:
Fix error in apptainer definition file in oddpub service

joshlawrimore committed:
debug oddpub definition.def file

joshlawrimore committed:
Updated permission of entrypoint in definition.def file

joshlawrimore committed:
Yet again fixing definition.def file

joshlawrimore committed:
Trying again with the definition.def file...

joshlawrimore committed:
Trying again with the definition.def file...

joshlawrimore committed:
Trying logging in the definition.def file...
